stream/stream_test.go

```diff
@@ -0,0 +1,139 @@
+				// and server-side CancelFuncs, so that parseStreamSpec
+				// can drive cancellation on either end of the Peer. To
+				// synchronize client-side cancellation properly, also
+				// smuggle in the client-side context, which will be used
+				// exclusively for synchronizing cancellation.
+				serverCtx, serverCancel := context.WithCancel(context.Background())
+				serverCtx = context.WithValue(serverCtx, serverCancelContextKey{}, serverCancel)
+				serverCtx = context.WithValue(serverCtx, clientCtxContextKey{}, ctx)
+				serverCtx = context.WithValue(serverCtx, clientCancelContextKey{}, clientCancel)
+				return serverCtx
+			})
+
+			var got []string
+			var gotErr error
+			for resp, err := range stream.Call(ctx, ps.A, "stream", []byte("req")) {
+				if err != nil {
+					gotErr = err
+					break
+				}
+				got = append(got, string(resp))
+			}
+			if !reflect.DeepEqual(got, tc.want) {
+				t.Fatalf("got stream %v, want %v", got, tc.want)
+			}
+			if gotErr != nil {
+				// Errors transit over a network, so we can't compare with
+				// errors.Is. Compare strings instead as an approximation.
+				if gotErr.Error() != tc.wantErr {
+					t.Fatalf("unexpected error %q, want %q", gotErr, tc.wantErr)
+				}
+			} else if tc.wantErr != "" {
+				t.Fatalf("stream didn't yield error, want %q", tc.wantErr)
+			}
+		})
+	}
+}
+
+func parseStreamSpec(t *testing.T, s string) stream.HandlerFunc {
+	return func(ctx context.Context, req *chirp.Request) iter.Seq2[[]byte, error] {
+		return func(yield func([]byte, error) bool) {
+			for _, cmd := range strings.Split(s, ",") {
+				fs := strings.Fields(cmd)
+				switch fs[0] {
+				case "stream":
+					for _, v := range fs[1:] {
+						if !yield([]byte(v), nil) {
+							return
+						}
+					}
+				case "req":
+					if !yield(req.Data, nil) {
+						return
+					}
+				case "err":
+					yield(nil, testErr)
+					return
+				case "server-cancel":
+					cancel := ctx.Value(serverCancelContextKey{}).(context.CancelFunc)
+					cancel()
+					// Closing of ctx.Done can happen asynchronously
+					// after cancel returns. Wait on ctx.Done
+					// ourselves, to ensure that the caller will
+					// reliably see a canceled context as well.
+					<-ctx.Done()
+				case "client-cancel":
+					cancel := ctx.Value(clientCancelContextKey{}).(context.CancelFunc)
+					cancel()
+					// Same as above, force synchronization of the
+					// client's context cancellation.
+					clientCtx := ctx.Value(clientCtxContextKey{}).(context.Context)
+					<-clientCtx.Done()
+				case "return-canceled":
+					if ctx.Err() == nil {
+						t.Errorf("parseStreamSpec instructed to return-canceled, but ctx isn't canceled")
+					}
+					yield(nil, ctx.Err())
+					return
+				default:
+					t.Errorf("unknown parseStreamSpec command %q", fs[0])
+				}
+			}
+		}
+	}
+}
+
+type clientCancelContextKey struct{}
+type clientCtxContextKey struct{}
+type serverCancelContextKey struct{}
+
+var testErr = errors.New("test")
```

Comment:
The linter is annoying, but probably worth doing.

Reply:
Alternatively, I think it won't do this check in local scopes, or you can add `//lint:ignore ST1012 For testing` at EOL. YMMV.